fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -885,6 +954,56 @@ protected void validateSameMetadataOrThrow(FDBRecordStore store) {
         }
     }
 
+    CompletableFuture<Void> indexingStamp(@Nonnull ConcurrentHashMap<String, IndexBuildProto.IndexBuildIndexingStamp> oldStamps,
+                                          @Nullable OnlineIndexer.IndexingStampOperation op,
+                                          @Nullable String id,
+                                          @Nullable  Integer timeoutSeconds) {
+        return getRunner().runAsync(context -> openRecordStore(context).thenCompose(store ->
+            forEachTargetIndex(index -> store.loadIndexingTypeStampAsync(index)
+                    .thenApply(stamp -> indexingStamp(oldStamps, store, index, stamp, op, id, timeoutSeconds)))
+        ));
+    }
+
+    boolean indexingStamp(@Nonnull ConcurrentHashMap<String, IndexBuildProto.IndexBuildIndexingStamp> oldStamps,
+                          @Nonnull FDBRecordStore store,
+                          @Nonnull Index index,
+                          @Nullable IndexBuildProto.IndexBuildIndexingStamp stamp,
+                          @Nullable OnlineIndexer.IndexingStampOperation op,
+                          @Nullable String id,
+                          @Nullable  Integer timeoutSeconds
+    ) {
+        oldStamps.put(index.getName(), stamp != null ? stamp :
+                                       IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                                               .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.INVALID)
+                                               .build());
+        if (op == null || stamp == null || op.equals(OnlineIndexer.IndexingStampOperation.QUERY)) {
+            return false;
```

Comment:
Even if you wanted to collect the results in parallel, it is probably better to return the map rather than have it as an argument, especially into the `indexingStamp` method that creates the transaction. If you want it to run in parallel, you can still do that by constructing the map within the `runAsync` lambda and then either passing it to this function or having this function return just the stamp for the index it works on, and then the results can be collected in one spot